### PR TITLE
Periodic Data Retrieval implementation

### PR DESCRIPTION
**Pull request by alex-deaconu**
Basic functionality is in the first commit.

Split out `cache.js` and `logger.js` (second commit).

Unit tests

### CODE REVIEW

**Comment by alex-deaconu** [comment]
@ezequielc please review. Thanks!

File: test/unit/rise-data-weather.html [hunk 1]
```diff
@@ -178,116 +152,315 @@
...
+        element._requestData();
+
+        setTimeout(() => {
+          assert.isFalse( element._sendWeatherEvent.called );
+
+          done();            
+        }, 10);
+      });
+
+      test( "should not send data-update if returned data has a different timestamp", done => {
```
Comment (ezequielc) [inline test/unit/rise-data-weather.html]: @alex-deaconu To confirm, this should say "should send data-update...", right?

File: src/rise-data-weather.js [hunk 2]
```diff
@@ -41,22 +44,22 @@ class RiseDataWeather extends PolymerElement {
...
 
-  static get CACHE_NAME() {
-    return "rise-data-weather";
-  }
-
-  static get CACHE_DURATION() {
-    return 1000 * 60 * 60 * 2;
+  static get FETCH_CONFIG() {
+    return {
+      INTERVAL: 1000 * 60,
```
Comment (ezequielc) [inline src/rise-data-weather.js]: @alex-deaconu For clarity, I think you should consider renaming this to RETRY_INTERVAL. Thanks

File: src/rise-data-weather.js [hunk 3]
```diff
@@ -203,47 +145,45 @@ class RiseDataWeather extends PolymerElement {
...
+
+      if ( !( this.weatherData && this.weatherData.reportDate ) || this.weatherData.reportDate.getTime() !== data.reportDate.getTime()) {
+        this._setWeatherData( data );
+
+        this._sendWeatherEvent( RiseDataWeather.EVENT_DATA_UPDATE, this.weatherData );
+      }
+
+      this._refresh( RiseDataWeather.FETCH_CONFIG.REFRESH );
     } catch ( e ) {
       this._sendWeatherEvent( RiseDataWeather.EVENT_DATA_ERROR, e );
```
Comment (ezequielc) [inline src/rise-data-weather.js]: @alex-deaconu In case of data-error, looks like there is no refresh. I think we should queue a refresh as it could be a temporary server error. Thanks

File: src/tinbu-parser.js [hunk 4]
```diff
@@ -27,8 +27,10 @@ function parseTinbu( body ) {
       .singleNodeValue,
 
     isMetric = report.getAttribute( "metric" ) === "true",
+    reportDate = new Date( report.getAttribute( "localtime" ) || Date.now()),
```
Comment (ezequielc) [inline src/tinbu-parser.js]: @alex-deaconu Any concerns with timezones here? Or both are UTC? Thanks

File: src/logger.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import { version } from "./rise-data-weather-version.js";
+
+var id = null;
+const logger = {
+  init,
+  log
+};
+
+function init( componentId ) {
```
Comment (ezequielc) [inline src/logger.js]: @alex-deaconu Will this work when there are multiple weather components on the same template? Looks like `id` is global. Thanks

File: test/unit/rise-data-weather.html (hunk 1 above)
Comment (alex-deaconu) [reply test/unit/rise-data-weather.html]: @ezequielc good catch, updated!

File: src/rise-data-weather.js (hunk 2 above)
Comment (alex-deaconu) [reply src/rise-data-weather.js]: @ezequielc good suggestion, updated!

File: src/tinbu-parser.js (hunk 4 above)
Comment (alex-deaconu) [reply src/tinbu-parser.js]: @ezequielc good question. First off, I was thinking to use the `localupdatetime` field. From what I can tell field is when the report was added. This would give us the best accuracy for this functionality as well as the cache refresh interval.

As for the timezone, I added `Date.now()` as a backup in case the field is missing in the XML to prevent errors in the comparison. Also, `localtime` has the timezone included so when converting to Date it should be ok. 

Overall, we are using this for comparison purposes to see if we should be sending `data-update` event, so in case of the fallback the event will be sent all the time.

In conclusion, I don't think we should worry about timezones, but maybe we should consider using `localupdatetime` here and for the cache comparison. Maybe we can queue that as an improvement. Thoughts? Thanks.

File: src/logger.js (hunk 5 above)
Comment (alex-deaconu) [reply src/logger.js]: @ezequielc good question. I would assume this variable is local to this particular module, but I will test with multiple components in the template just to make sure!

File: src/rise-data-weather.js (hunk 3 above)
Comment (alex-deaconu) [reply src/rise-data-weather.js]: @ezequielc good point. Should we send the event after each failure or only after failing 5 retries? Thanks!

File: src/rise-data-weather.js (hunk 2 above)
Comment (ezequielc) [reply src/rise-data-weather.js]: Thanks!

File: src/rise-data-weather.js (hunk 3 above)
Comment (ezequielc) [reply src/rise-data-weather.js]: @alex-deaconu I was thinking about something simple here: send the data-error event and queue a new refresh after 60 minutes.

It could be something wrong with the server, like 500 error, but it could also be an invalid city, so I don't think we need to keep retrying very often. Thanks

File: src/tinbu-parser.js (hunk 4 above)
Comment (ezequielc) [reply src/tinbu-parser.js]: @alex-deaconu Ok, I think we can queue it as an improvement. And given we increased the refresh interval, this is no longer as important as it was initially. Thanks!

File: src/logger.js (hunk 5 above)
Comment (ezequielc) [reply src/logger.js]: @alex-deaconu Ok, let me know the results. Thanks

Comment (alex-deaconu) [reply src/logger.js]: @ezequielc silly assumption. Will fix. Thanks!

File: src/rise-data-weather.js [hunk 6]
```diff
@@ -203,47 +145,45 @@ class RiseDataWeather extends PolymerElement {
...
     fetch( this._getUrl(), {
       headers: { "X-Requested-With": "rise-data-weather" }
     }).then( res => {
-      return this._getCache().then( cache => {
-        this._handleResponse( res.clone());
-        return cache.put( res.url, res );
-      })
+      this._handleResponse( res.clone());
+
+      cache.put( res );
```
Comment (alex-deaconu) [inline src/rise-data-weather.js]: @ezequielc there was a typo here which was causing caching errors. Also, the error was returned here and was handled by `_handleFetchError` which was causing error retries. Corrected both in the latest commit, wondering if you could review. Thanks!

Comment (ezequielc) [reply src/rise-data-weather.js]: @alex-deaconu Good catch. Correction looks good.
